book-content/chapters/02-ide-superpowers.md

```diff
@@ -0,0 +1,579 @@
+```
+
+This tells us that the `title` property in the `job` object is the problem.
+
+Already, we understand the issue without needing to see the top line, which is usually a long summary of the problem.
+
+Reading errors bottom-to-top can be a helpful strategy when dealing with multi-line TypeScript errors.
+
+## Introspecting Variables and Declarations
+
+You can hover over more than just error messages. Any time you hover over a variable or declaration, VS Code will show you information about it.
+
+In this example, we could hover over `thing` and see that it's of type `number`:
+
+```ts twoslash
+let thing = 123;
+//  ^?
+```
+
+Hovering works for more involved examples as well. Here `otherObject` spreads in the properties of `otherThing` as well as adding `thing`:
+
+```typescript
+let otherThing = {
+  name: "Alice",
+};
+
+const otherObject = {
+  ...otherThing,
+  thing: "abc",
+};
+
+otherObject.thing;
+```
+
+Hovering over `otherObject` will give us a computed readout of all of its properties:
+
+```ts twoslash
+let otherThing = {
+  name: "Alice",
+};
+
+const otherObject = {
+  ...otherThing,
+  thing: "abc",
+};
+
+// ---cut---
+console.log(otherObject);
+//          ^?
+```
+
+Depending on what you hover over, VS Code will show you different information. For example, hovering over `otherObject` will show you all of its properties, while hovering over `thing` will show you its type.
+
+Get used to the ability to float around your codebase introspecting variables and declarations, because it's a great way to understand what the code is doing.
+
+### Exercises
+
+#### Exercise 1: Hovering a Function Call
+
+In this code snippet we're trying to grab an element using `document.getElementById` with an ID of `12`. However, TypeScript is complaining.
+
+```ts twoslash
+// @errors: 2345
+let element = document.getElementById(12);
+```
+
+How can hovering help to determine what argument `document.getElementById` actually requires? And for a bonus point, what type is `element`?
+
+<Exercise title="Exercise 1: Hovering a Function Call" filePath="/src/016.5-ide-superpowers/041-hovering-a-function-call.problem.ts"></Exercise>
+
+#### Solution 1: Hovering a Function Call
+
+First of all, we can hover over the red squiggly error itself. In this case, hovering over `12`, we get the following error message:
+
+```
+Argument of type 'number' is not assignable to parameter of type 'string'.
+```
+
+We'll also get a readout of the `getElementById` function:
+
+```
+(method) Document.getElementById(elementId: string): HTMLElement | null
+```
+
+In the case of `getElementById`, we can see that it requires a string as an argument, and it returns an `HTMLElement | null`. We'll look at this syntax later, but it basically means either a `HTMLElement` or `null`.
+
+This tells us that we can fix the error by changing the argument to a string:
+
+```ts twoslash
+let element = document.getElementById("12");
```

Comment:
`"12"` is technically a valid ID... but it just looks so weird-looking. I'd try to pick a different value here. I don't think it matters that this is what the user tried to pass before but quoted.